conda/notices/core.py

```diff
@@ -35,69 +36,104 @@ def display_notices(
+    cache_file = cache.get_notices_cache_file()
+    cache.mark_channel_notices_as_viewed(cache_file, channel_notice_set.channel_notices)
 
-    views.print_more_notices_message(num_total_notices, len(channel_notices), num_viewed_notices)
+    views.print_more_notices_message(
+        channel_notice_set.total_number_channel_notices,
+        len(channel_notice_set.channel_notices),
+        channel_notice_set.viewed_channel_notices,
+    )
 
 
 def notices(func):
     """
     Wrapper for "execute" entry points for subcommands.
 
+    If channel notices need to be fetched, we do that first and then
+    run the command normally. We then display these notices at the very
+    end of the command output so that the user is more likely to see them.
+
+    This ordering was specifically done to address the following bug report:
+        - https://github.com/conda/conda/issues/11847
+
     Args:
         func: Function to be decorated
     """
-
     @wraps(func)
     def wrapper(*args, **kwargs):
-        return_value = func(*args, **kwargs)
-
-        if is_channel_notices_enabled(context):
-            display_notices(
+        if is_channel_notices_enabled(context) and is_channel_notices_cache_expired():
+            channel_notice_set = retrieve_notices(
                 limit=context.number_channel_notices,
                 always_show_viewed=False,
                 silent=True,
             )
+            return_value = func(*args, **kwargs)
+            display_notices(channel_notice_set)
+
+            return return_value
 
-        return return_value
+        else:
+            return func(*args, **kwargs)
 
     return wrapper
 
 
 def get_channel_name_and_urls(
-    channels: [Sequence[Channel]],
-) -> Sequence[Tuple[ChannelUrl, ChannelName]]:
-    """Return a sequence of Channel URL and name"""
-
-    def ensure_endswith(value: str, ends: str) -> str:
-        return value if value.endswith(ends) else f"{value}{ends}"
+    channels: Sequence[Union[Channel, MultiChannel]],
+) -> Sequence[tuple[ChannelUrl, ChannelName]]:
+    """
+    Return a sequence of Channel URL and name tuples.
 
+    This function handles both Channel and MultiChannel object types.
+    """
     def join_url(value: str, join_val: str) -> str:
-        return parse.urljoin(ensure_endswith(value, "/"), join_val)
+        """Using this method because urllib.parse.urljoin does not support s3:// URLs"""
+        ending = "/"
+        value = value if value.endswith(ending) else f"{value}{ending}"
+        return f"{value}{join_val}"
 
-    return tuple(
-        (join_url(base_url, NOTICES_FN), channel.name or channel.location)
-        for channel in channels
-        for base_url in channel.base_urls
-    )
+    channel_name_and_urls = []
+
+    for channel in channels:
+        name = channel.name or channel.location
+
+        for url in channel.base_urls:
+            channel_name_and_urls.append((join_url(url, NOTICES_FN), name))
```

Comment:
We can add the s3 protocol such that `urllib.parse.urljoin` will work:

```
urllib.parse.uses_relative.append("s3")
urllib.parse.uses_netloc.append("s3")
```

Alternatively, I'd suggest we inline the helper function:

```suggestion
            channel_name_and_urls.append((f"{url.rstrip('/')}/{NOTICES_FN}", name))
```

Reply:
I choose to get rid of the urllib.parse modification because it felt a little hacky. You don't think so?

Reply:
Ah, I was trying to figure out where I'd seen this before, in that case, I'd just go with inlining this

Reply:
Will do. It's not going to be as inline as you have it because I really don't like having two levels of quotes (one with `"` and the other with `'`).